clang/test/Analysis/Checkers/WebKit/call-args.cpp

```diff
@@ -365,3 +365,20 @@ namespace call_with_explicit_temporary_obj {
     RefPtr { provide() }->method();
   }
 }
+
+namespace call_with_adopt_ref {
+  class Obj {
+  public:
+    void ref() const;
+    void deref() const;
+    void method();
+  };
+
+  struct dummy {
+    RefPtr<Obj> any;
```

Comment:
For whatever reason, this must exist for the bug to reproduce.

Reply:
@rniwa I ran this test case under the debugger and here's what I found: The call to `IsPtrOriginSafe` returns `true` in the absence of `struct dummy` when visiting the `CallExpr` causing the checker to not issue a warning:
```
CXXMemberCallExpr 0x147945af8 'void'
`-MemberExpr 0x147945ac8 '<bound member function type>' ->method 0x14793ccb0
  `-CXXOperatorCallExpr 0x147945988 'class call_with_adopt_ref::Obj *' '->'
    |-ImplicitCastExpr 0x147945970 'class call_with_adopt_ref::Obj *(*)(void) const' <FunctionToPointerDecay>
    | `-DeclRefExpr 0x1479458e8 'class call_with_adopt_ref::Obj *(void) const' lvalue CXXMethod 0x147944388 'operator->' 'class call_with_adopt_ref::Obj *(void) const'
    `-ImplicitCastExpr 0x1479458d0 'const struct RefPtr<class call_with_adopt_ref::Obj>' lvalue <NoOp>
      `-MaterializeTemporaryExpr 0x1479458b8 'RefPtr<Obj>':'struct RefPtr<class call_with_adopt_ref::Obj>' lvalue
        `-CXXBindTemporaryExpr 0x147945898 'RefPtr<Obj>':'struct RefPtr<class call_with_adopt_ref::Obj>' (CXXTemporary 0x147945898)
          `-CallExpr 0x1479407e8 'RefPtr<Obj>':'struct RefPtr<class call_with_adopt_ref::Obj>'
            |-ImplicitCastExpr 0x1479407d0 'RefPtr<Obj> (*)(class call_with_adopt_ref::Obj *)' <FunctionToPointerDecay>
            | `-DeclRefExpr 0x147940740 'RefPtr<Obj> (class call_with_adopt_ref::Obj *)' lvalue Function 0x14793fe00 'adoptRef' 'RefPtr<Obj> (class call_with_adopt_ref::Obj *)' (FunctionTemplate 0x14792ff48 'adoptRef')
            `-CXXNewExpr 0x14793f408 'Obj *' Function 0x14793d0c8 'operator new' 'void *(unsigned long)'
              `-CXXConstructExpr 0x14793f3e0 'Obj':'class call_with_adopt_ref::Obj' 'void (void) noexcept' 
```

Does this give any clue for the unexpected behavior?

Reply:
It's still unclear to me why the presence of struct dummy affects the semantics of seemingly unrelated code.

Reply:
Looks like there is some bug in AST generation. The return type of callee is different for both the versions at
 https://github.com/llvm/llvm-project/blob/fbc18b85d6ce5ab6489a2b08f9b38d446fe9d6f6/clang/lib/StaticAnalyzer/Checkers/WebKit/ASTUtils.cpp#L103

With dummy struct:
```
(lldb) p callee->getReturnType()->dump()
ElaboratedType 0x1309723b0 'RefPtr<Obj>' sugar
`-RecordType 0x130971240 'struct RefPtr<class call_with_adopt_ref::Obj>'
  `-ClassTemplateSpecialization 0x130971160 'RefPtr'

```
Without the dummy struct:
```
(lldb) p callee->getReturnType()->dump()
ElaboratedType 0x12c973230 'RefPtr<Obj>' sugar
`-TemplateSpecializationType 0x12c9731e0 'RefPtr<class call_with_adopt_ref::Obj>' sugar
  |-name: 'RefPtr' qualified
  | `-ClassTemplateDecl 0x12c95ee30 prev 0x12c95e690  RefPtr
  |-TemplateArgument type 'class call_with_adopt_ref::Obj'
  | `-SubstTemplateTypeParmType 0x12c9730a0 'class call_with_adopt_ref::Obj' sugar typename depth 0 index 0 T
  |   |-FunctionTemplate 0x12c95ec18 'adoptRef'
  |   `-RecordType 0x12c970560 'class call_with_adopt_ref::Obj'
  |     `-CXXRecord 0x12c9704d0 'Obj'
  `-RecordType 0x12c9731c0 'struct RefPtr<class call_with_adopt_ref::Obj>'
    `-ClassTemplateSpecialization 0x12c9730e0 'RefPtr'
```

Maybe create a separate radar for this issue and add the radar ID to the comment in the test case?
Otherwise, LGTM!

Reply:
Filed rdar://141692212. Will add a comment to that end. Thanks for the review!